ui/StatusQ/src/wallet/managetokenscontroller.cpp

```diff
@@ -0,0 +1,364 @@
+    rebuildCommunityTokenGroupsModel();
+}
+
+void ManageTokensController::saveSettings()
+{
+    Q_ASSERT(!m_settingsKey.isEmpty());
+
+    // gather the data to save
+    SerializedTokenData result;
+    for (auto model: {m_regularTokensModel, m_communityTokensModel})
+        result.insert(model->save());
+    result.insert(m_hiddenTokensModel->save(false));
+
+    // save to QSettings
+    m_settings.beginGroup(QStringLiteral("ManageTokens-%1").arg(m_settingsKey));
+    m_settings.beginWriteArray(m_settingsKey);
+    SerializedTokenData::const_key_value_iterator it = result.constKeyValueBegin();
+    for (auto i = 0; it != result.constKeyValueEnd() && i < result.size(); it++, i++) {
+        m_settings.setArrayIndex(i);
+        const auto tuple = it->second;
+        m_settings.setValue(QStringLiteral("symbol"), it->first);
+        m_settings.setValue(QStringLiteral("pos"), std::get<0>(tuple));
+        m_settings.setValue(QStringLiteral("visible"), std::get<1>(tuple));
+        m_settings.setValue(QStringLiteral("groupId"), std::get<2>(tuple));
+    }
+    m_settings.endArray();
+    m_settings.endGroup();
+    m_settings.sync();
+
+    // unset dirty
+    for (auto model: m_allModels)
+        model->setDirty(false);
+}
+
+void ManageTokensController::clearSettings()
+{
+    Q_ASSERT(!m_settingsKey.isEmpty());
+
+    // load from QSettings
+    m_settings.beginGroup(QStringLiteral("ManageTokens-%1").arg(m_settingsKey));
+    m_settings.remove(QString());
+    m_settings.endGroup();
+    m_settings.sync();
+}
+
+void ManageTokensController::loadSettings()
+{
+    Q_ASSERT(!m_settingsKey.isEmpty());
+
+    m_settingsData.clear();
+
+    m_settings.beginGroup(QStringLiteral("ManageTokens-%1").arg(m_settingsKey));
+    const auto size = m_settings.beginReadArray(m_settingsKey);
+    for (auto i = 0; i < size; i++) {
+        m_settings.setArrayIndex(i);
+        const auto symbol = m_settings.value(QStringLiteral("symbol")).toString();
+        if (symbol.isEmpty()) {
+            qCWarning(manageTokens) << Q_FUNC_INFO << "Missing symbol while reading tokens settings";
+            continue;
+        }
+        const auto pos = m_settings.value(QStringLiteral("pos"), -1).toInt();
+        const auto visible = m_settings.value(QStringLiteral("visible"), true).toBool();
+        const auto groupId = m_settings.value(QStringLiteral("groupId")).toString();
+        m_settingsData.insert(symbol, {pos, visible, groupId});
+    }
+    m_settings.endArray();
+    m_settings.endGroup();
+}
+
+void ManageTokensController::revert()
+{
+    loadSettings();
+    parseSourceModel();
+}
+
+void ManageTokensController::classBegin()
+{
+    // empty on purpose
+}
+
+void ManageTokensController::componentComplete()
+{
+    loadSettings();
+}
+
+void ManageTokensController::setSourceModel(QAbstractItemModel* newSourceModel)
+{
+    if(!newSourceModel) return;
+    if(m_sourceModel == newSourceModel) return;
+    m_sourceModel = newSourceModel;
```

Comment:
```
    if(!newSourceModel) {
         disconnect(sourceModel());
         return;
     }
```

Sorry for buzzing you with this, but I think the newly introduced disconnect has two disadvantages in this form:
1. It's not setting the `m_sourceModel` to null when `setSourceModel(nullptr)` is called As a result the `revert` flow becomes unstable an could easily crash the app because it will dereference the old `m_sourceModel`.
2. Doesn't disconnect from the old model if the `newSourceModel` is valid. As a result `ManageTokensController` will be connected to both models on these events: `rowsInserted`, `rowsRemoved`, `dataChanged`